File: src/event.cpp
#include <algorithm>
#include <random>

#include "event.hpp"

Event::Event(std::chrono::milliseconds onset,
             std::chrono::milliseconds duration, std::string trial_type)
    : onset_(onset), duration_(duration), trial_type_(trial_type) {}

std::chrono::milliseconds Event::onset() const { return onset_; }

std::chrono::milliseconds Event::duration() const { return duration_; }

std::string const& Event::trial_type() const { return trial_type_; }

std::vector<int> Event::sample_onsets(std::vector<Event> const& events,
                                      int sfreq) {
  std::vector<int> sample_onsets;
  sample_onsets.reserve(events.size());

  for (auto const& event : events) {
    int time_onset = static_cast<int>(event.onset().count());
    sample_onsets.push_back(time_onset * sfreq);
  }

  return sample_onsets;
}

std::vector<Event> Event::generate_fixed_duration_events(
    std::chrono::milliseconds duration,
    std::vector<std::string> const& trial_types,
    std::chrono::milliseconds task_duration) {
  int n_events = static_cast<int>(task_duration.count() / duration.count());
  std::vector<Event> events;
  events.reserve(n_events);

  for (int i = 0; i < n_events; ++i) {
    std::chrono::milliseconds onset(i * duration.count());
    std::string trial_type = trial_types[i % trial_types.size()];
    events.emplace_back(onset, duration, trial_type);
  }

  return events;
}

std::vector<Event> Event::generate_variable_duration_events(
    std::chrono::milliseconds min_duration,
    std::chrono::milliseconds max_duration,
    std::vector<std::string> const& trial_types,
    std::chrono::milliseconds task_duration) {
  int max_n_events =
      static_cast<int>(task_duration.count() / min_duration.count());
  std::vector<double> event_durations_array(max_n_events);
  std::vector<std::chrono::milliseconds> event_durations;
  event_durations.reserve(max_n_events);

  std::random_device rd;
  std::mt19937 generator(rd());
  std::uniform_real_distribution<double> distribution(0.0, 1.0);
  double shift = static_cast<double>(min_duration.count());
  double scale =
      static_cast<double>(max_duration.count() - min_duration.count());

  for (auto& duration : event_durations_array) {
    duration = distribution(generator) * scale + shift;
    event_durations.push_back(
        std::chrono::milliseconds(static_cast<long long>(duration)));
  }

  std::vector<std::chrono::milliseconds> event_onsets;
  event_onsets.reserve(max_n_events);
  std::vector<double> event_onsets_array = {0.0};
  std::partial_sum(event_durations.cbegin(), event_durations.cend(),
                   std::back_inserter(event_onsets));

  int n_types = static_cast<int>(trial_types.size());
  int n_repeats =
      static_cast<int>(std::ceil(static_cast<double>(max_n_events) / n_types));
  std::vector<std::string> event_types;
  event_types.reserve(max_n_events);

  if (n_types > 2) {
    // Create a temporary list of events
    std::vector<std::vector<std::string>> list_events;
    list_events.reserve(n_repeats);

    // type_set = a set of trial types
    std::vector<std::string> type_set(trial_types.begin(), trial_types.end());
    std::shuffle(type_set.begin(), type_set.end(), std::random_device());
    list_events.push_back(type_set);

    for (int i = 0; i < n_repeats - 1; ++i) {
      type_set.clear();
      type_set.assign(trial_types.begin(), trial_types.end());
      std::shuffle(type_set.begin(), type_set.end(), std::random_device());

      while (type_set[0] == list_events.back().back()) {
        type_set.clear();
        type_set.assign(trial_types.begin(), trial_types.end());
        std::shuffle(type_set.begin(), type_set.end(), std::random_device());
      }

      list_events.push_back(type_set);
    }

    for (auto const& event_set : list_events) {
      int n_events_remaining = max_n_events - event_types.size();
      if (n_events_remaining < event_set.size()) {
        int n_trunc = event_set.size() - n_events_remaining;
        event_types.insert(event_types.end(), event_set.begin(),
                           event_set.end() - n_trunc);
        break;
      } else {
        event_types.insert(event_types.end(), event_set.begin(),
                           event_set.end());
      }
    }
  } else {
    event_types.reserve(n_repeats * n_types);

    for (int i = 0; i < n_repeats; ++i) {
      for (auto const& trial_type : trial_types) {
        event_types.push_back(trial_type);
      }
    }
  }

  std::vector<Event> events;
  events.reserve(max_n_events);

  for (int i = 0; i < max_n_events; ++i) {
    std::chrono::milliseconds onset = event_onsets[i];
    std::chrono::milliseconds duration = event_durations[i];
    std::string trial_type = event_types[i];
    events.emplace_back(onset, duration, trial_type);
  }

  return events;
}

File: libbids/event.py
import threading
import numpy as np
from datetime import timedelta
from typing import (
    Any,
    Dict,
    Iterator,
    List,
    Optional,
    Union,
    cast,
)


class Event:
    def __init__(
        self,
        onset: Optional[Union[timedelta, float]] = None,
        duration: Optional[Union[timedelta, float]] = None,
        trial_type: Optional[str] = None,
        triggerable: bool = False,
    ):
        """Create an event to be recorded or delivered during a task

        Parameters
        ----------
        onset : Optional[Union[timedelta, float]]
            The time since the begining of the run until this event
        duration : Optional[Union[timedelta, float]]
            The duration of the event
        trial_type : Optional[str]
            The event label
        triggerable : bool
            Whether the event can be stopped prematurely; i.e., duration can be
            shortened based on experimental events in real-time
        """
        self.onset: Optional[timedelta] = (
            timedelta(seconds=cast(float, onset))
            if type(onset) not in [type(None), timedelta]
            else cast(Optional[timedelta], onset)
        )
        self.duration: Optional[timedelta] = (
            timedelta(seconds=cast(float, duration))
            if type(duration) not in [type(None), timedelta]
            else cast(Optional[timedelta], duration)
        )
        self.trial_type: Optional[str] = trial_type
        self.triggerable: bool = triggerable
        self.threading_event: threading.Event = threading.Event()

    def __iter__(self) -> Iterator:
        keys: List[str] = ["onset", "duration", "trial_type"]
        for key in keys:
            value = self.__getattribute__(key)
            if isinstance(value, timedelta):
                yield (key, value.total_seconds())
            else:
                yield (key, value)

    def __repr__(self) -> str:
        return str(self.to_dict())

    def is_set(self) -> bool:
        return self.threading_event.is_set()

    def set(self) -> None:
        if self.triggerable:
            self.threading_event.set()

    def to_dict(self) -> Dict[str, Any]:
        return {k: v for k, v in self}

    @property
    def is_timed(self) -> bool:
        return (self.onset is not None) and (self.duration is not None)

    @classmethod
    def sample_onsets(cls, events: List["Event"], sfreq: int) -> np.ndarray:
        """Generate an array of sample indices that correlate to the onset of
        each event listed in events

        Parameters
        ----------
        events : List[Event]
            List of events to convert to sample onsets
        sfreq : int
            The sampling frequency used to convert from seconds to samples

        Returns
        -------
        np.ndarray
            The array of sample onsets
        """
        time_onsets: np.ndarray = np.array(
            [cast(timedelta, e.onset).total_seconds() for e in events]
        )
        return time_onsets * sfreq

    @classmethod
    def generate_fixed_duration_events(
        cls, duration: timedelta, trial_types: List[str], task_duration: timedelta
    ) -> List["Event"]:
        """Generates a list of events. Each event will have a fixed width. The
        number of events returned will be enough to fill up the `task_duration`

        Parameters
        ----------
        duration : timedelta
            The duration that each event should last
        trial_types : List[str]
            A unique list of event names.
        task_duration : timedelta
            The total duration of the task

        Returns
        -------
        List[Event]
            A list of events that will define the task, or part of a task
        """
        n_events: int = int(task_duration.total_seconds() // duration.total_seconds())
        return [
            Event(
                onset=timedelta(seconds=i * duration.total_seconds()),
                duration=duration,
                trial_type=trial_types[i % len(trial_types)],
            )
            for i in range(n_events)
        ]

    @classmethod
    def generate_variable_duration_events(
        cls,
        min_duration: timedelta,
        max_duration: timedelta,
        trial_types: List[str],
        task_duration: timedelta,
    ) -> List["Event"]:
        max_n_events: int = int(
            task_duration.total_seconds() // min_duration.total_seconds()
        )
        shift: float = min_duration.total_seconds()
        scale: float = max_duration.total_seconds() - min_duration.total_seconds()
        event_durations_array: np.ndarray = np.random.rand(max_n_events) * scale + shift
        event_durations: List[timedelta] = [
            timedelta(seconds=s) for s in event_durations_array
        ]
        event_onsets_array: np.ndarray = [0] + np.cumsum(
            [i.total_seconds() for i in event_durations]
        ).tolist()[:-1]
        event_onsets: List[timedelta] = [
            timedelta(seconds=d) for d in event_onsets_array
        ]
        n_types: int = len(trial_types)
        n_repeats: int = int(np.ceil(max_n_events / n_types))
        if n_types > 2:
            list_events: List = []
            type_set: np.ndarray = np.random.default_rng().choice(
                trial_types, size=n_types, replace=False
            )
            list_events.append(type_set.tolist())
            for _ in range(n_repeats - 1):
                type_set = np.random.default_rng().choice(
                    trial_types, size=n_types, replace=False
                )
                # Loop ensures that types do not immediately repeat
                while type_set[0] == list_events[-1][-1]:
                    type_set = np.random.default_rng().choice(
                        trial_types, size=n_types, replace=False
                    )
                list_events.append(type_set.tolist())
            event_types: np.ndarray = np.hstack(list_events)
        else:
            event_types = np.array(trial_types * n_repeats)
        return [
            Event(onset, duration, trial_type)
            for onset, duration, trial_type in zip(
                event_onsets, event_durations, event_types
            )
        ]
